tests/test-spawn.lua

```diff
@@ -321,19 +319,47 @@ local steps = {
         return true
     end,
     -- Test the matcher
+    function()
+        -- pre spawn client
+        if #client.get() ~= 0 then return end
+        spawn(tiny_client("client1"))
+
+        return true
+    end,
+    function()
+        -- test matcher
+        if #client.get() ~= 1 then return end
+
+        local matcher = function(c)
+            return c.class == "client1" or c.class == "client2"
+        end
+
+        -- This should do nothing
+        spawn.once(tiny_client("client2"), {tag=screen[1].tags[5]}, matcher)
+        spawn.single_instance(tiny_client("client2"), {tag=screen[1].tags[5]}, matcher)
+
+        return true
+    end,
+    function()
+        -- clean up
+        if #client.get() ~= 1 then return end
+        client.get()[1]:kill()
+        return true
+    end,
+    -- Test rules works with matcher if client doesn't support startup id
+    -- Can this test be performed using '_client' module without external 'xterm' application?
```

Comment:
> -- Can this test be performed using '_client' module without external 'xterm' application?

`_client` support startup notification, which allows properties to be applied without a matcher. `xterm` doesn't support this. So, in practice they will always go down 2 different code paths. In the perfect world, we could use another tiny X11 client with the same-ish size_hints support and get rid of xterm as a test dependency.

What has to be tested:

 * `_client` without a matcher (using startup notifications)
 * `_client` with a matcher to make sure the matcher is used
 * `xterm` with a matcher, because it wont work without one

Reply:
If the `xterm` dependency is a problem, we could always just compile our own test program and use that. `test-gravity.c` already does this.

Reply:
@Elv13 
The question was can `_client` (or its mod) be used  as client without startup notification. I saw [this test](https://github.com/awesomeWM/awesome/blob/master/tests/test-spawn-snid.lua#L33) and suggest that some build-in gtk3 lgi app can emulate `xterm` behavior. But possible I get it wrong.

@psychon 
Not a problem at all, but would be nice get rid of external applications for tests. I personally had no `xterm` nor `xeyes` install. There are some build-in apps done with lgi so having `xterm` extra feels like half done work.  Still not sure if it's worth it.